protocol.go

```diff
@@ -48,6 +48,47 @@ const (
 
 var errNoTimeout = errors.New("no timeout")
 
+// ProtocolFromRequest returns the inferred protocol name for parsing an
+// HTTP request. It inspects the request's method and headers to determine the
+// protocol. If the request doesn't match any known protocol, an empty string
+// and false is returned.
+func ProtocolFromRequest(request *http.Request) (string, bool) {
```

Comment:
I've been thinking about if we'd want to ever return something more structured/richer -- like a `Protocol` type that can inform the caller if it's connect steram vs. connect unary or even the version (in the event we ever add a v2 of the protocol).

I suppose we can leave this as is. In `Spec`, we'd have to add another accessor -- like maybe `ProtocolDetails`, and maybe we could then deprecate this and add `ProtocolDetailsFromRequest` at that time. WDYT?

Reply:
> My preference would be to move this back to authn-go until we comeup with a better implementation.

It would be nice to have some canonical implementation that we can also use from vanguard. Admittedly, in the past, we (with @akshayjshah) discussed a possible `connectrpc/protocol-go` repo where we could put lower-level details of the protocol like this, and `connectrpc/connect-go` (et al) could then import that. I'm a little concerned with putting it in authn-go if we expect to move it out later. But I guess it's okay as long as we have this figured out before we get to a v1.0 of authn-go 🤷.